// cdktf/lib/backends/remote-backend.js
"use strict";
var _a, _b, _c, _d;
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrefixedRemoteWorkspaces = exports.NamedRemoteWorkspace = exports.DataTerraformRemoteState = exports.RemoteBackend = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const terraform_backend_1 = require("../terraform-backend");
const util_1 = require("../util");
const terraform_remote_state_1 = require("../terraform-remote-state");
/**
 * @experimental
 */
class RemoteBackend extends terraform_backend_1.TerraformBackend {
    /**
     * @experimental
     */
    constructor(scope, props) {
        super(scope, "backend", "remote");
        this.props = props;
    }
    /**
     * @experimental
     */
    synthesizeAttributes() {
        return util_1.keysToSnakeCase({ ...this.props });
    }
}
exports.RemoteBackend = RemoteBackend;
_a = JSII_RTTI_SYMBOL_1;
RemoteBackend[_a] = { fqn: "cdktf.RemoteBackend", version: "0.4.1" };
/**
 * @experimental
 */
class DataTerraformRemoteState extends terraform_remote_state_1.TerraformRemoteState {
    /**
     * @experimental
     */
    constructor(scope, id, config) {
        super(scope, id, "remote", config);
    }
}
exports.DataTerraformRemoteState = DataTerraformRemoteState;
_b = JSII_RTTI_SYMBOL_1;
DataTerraformRemoteState[_b] = { fqn: "cdktf.DataTerraformRemoteState", version: "0.4.1" };
/**
 * @experimental
 */
class NamedRemoteWorkspace {
    /**
     * @experimental
     */
    constructor(name) {
        this.name = name;
    }
}
exports.NamedRemoteWorkspace = NamedRemoteWorkspace;
_c = JSII_RTTI_SYMBOL_1;
NamedRemoteWorkspace[_c] = { fqn: "cdktf.NamedRemoteWorkspace", version: "0.4.1" };
/**
 * @experimental
 */
class PrefixedRemoteWorkspaces {
    /**
     * @experimental
     */
    constructor(prefix) {
        this.prefix = prefix;
    }
}
exports.PrefixedRemoteWorkspaces = PrefixedRemoteWorkspaces;
_d = JSII_RTTI_SYMBOL_1;
PrefixedRemoteWorkspaces[_d] = { fqn: "cdktf.PrefixedRemoteWorkspaces", version: "0.4.1" };
