// cdktf/lib/app.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.App = exports.CONTEXT_ENV = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const constructs_1 = require("constructs");
const fs = require("fs");
const package_json_1 = require("../package.json");
const manifest_1 = require("./manifest");
exports.CONTEXT_ENV = 'CDKTF_CONTEXT_JSON';
/**
 * (experimental) Represents a cdktf application.
 *
 * @experimental
 */
class App extends constructs_1.Construct {
    /**
     * (experimental) Defines an app.
     *
     * @param options configuration options.
     * @experimental
     */
    constructor(options = {}) {
        var _b, _c;
        super(undefined, '');
        this.outdir = (_c = (_b = process.env.CDKTF_OUTDIR) !== null && _b !== void 0 ? _b : options.outdir) !== null && _c !== void 0 ? _c : 'cdktf.out';
        this.targetStackId = process.env.CDKTF_TARGET_STACK_ID;
        this.loadContext(options.context);
        const node = constructs_1.Node.of(this);
        if (options.stackTraces === false) {
            node.setContext(constructs_1.ConstructMetadata.DISABLE_STACK_TRACE_IN_METADATA, true);
        }
        node.setContext('cdktfVersion', package_json_1.version);
    }
    /**
     * (experimental) Synthesizes all resources to the output directory.
     *
     * @experimental
     */
    synth() {
        if (!fs.existsSync(this.outdir)) {
            fs.mkdirSync(this.outdir);
        }
        const manifest = new manifest_1.Manifest(package_json_1.version, this.outdir);
        constructs_1.Node.of(this).synthesize({
            outdir: this.outdir,
            sessionContext: {
                manifest
            }
        });
        manifest.writeToFile();
    }
    loadContext(defaults = {}) {
        const node = constructs_1.Node.of(this);
        // prime with defaults passed through constructor
        for (const [k, v] of Object.entries(defaults)) {
            node.setContext(k, v);
        }
        // read from environment
        const contextJson = process.env[exports.CONTEXT_ENV];
        const contextFromEnvironment = contextJson
            ? JSON.parse(contextJson)
            : {};
        for (const [k, v] of Object.entries(contextFromEnvironment)) {
            node.setContext(k, v);
        }
    }
}
exports.App = App;
_a = JSII_RTTI_SYMBOL_1;
App[_a] = { fqn: "cdktf.App", version: "0.4.1" };
