// cdktf/lib/manifest.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Manifest = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const constructs_1 = require("constructs");
const path = require("path");
const fs = require("fs");
/**
 * @experimental
 */
class Manifest {
    /**
     * @experimental
     */
    constructor(version, outdir) {
        this.version = version;
        this.outdir = outdir;
        /**
         * @experimental
         */
        this.stacks = [];
        const stacksPath = path.join(this.outdir, Manifest.stacksFolder);
        if (!fs.existsSync(stacksPath))
            fs.mkdirSync(stacksPath);
    }
    /**
     * @experimental
     */
    forStack(stack) {
        const node = constructs_1.Node.of(stack);
        const manifest = {
            name: node.id,
            constructPath: node.path,
            workingDirectory: path.join(Manifest.stacksFolder, node.id),
            synthesizedStackPath: path.join(Manifest.stacksFolder, node.id, Manifest.stackFileName)
        };
        this.stacks.push(manifest);
        return manifest;
    }
    /**
     * @experimental
     */
    buildManifest() {
        return {
            version: this.version,
            stacks: this.stacks.reduce((newObject, stack) => {
                newObject[stack.name] = stack;
                return newObject;
            }, {})
        };
    }
    /**
     * @experimental
     */
    writeToFile() {
        fs.writeFileSync(path.join(this.outdir, Manifest.fileName), JSON.stringify(this.buildManifest(), undefined, 2));
    }
}
exports.Manifest = Manifest;
_a = JSII_RTTI_SYMBOL_1;
Manifest[_a] = { fqn: "cdktf.Manifest", version: "0.4.1" };
/**
 * @experimental
 */
Manifest.fileName = 'manifest.json';
/**
 * @experimental
 */
Manifest.stacksFolder = 'stacks';
/**
 * @experimental
 */
Manifest.stackFileName = 'cdk.tf.json';
