// cdktf/lib/terraform-element.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.TerraformElement = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const constructs_1 = require("constructs");
const terraform_stack_1 = require("./terraform-stack");
/**
 * @experimental
 */
class TerraformElement extends constructs_1.Construct {
    /**
     * @experimental
     */
    constructor(scope, id) {
        super(scope, id);
        /**
         * @experimental
         */
        this.rawOverrides = {};
        this.constructNode.addMetadata('stacktrace', 'trace');
        this.cdktfStack = terraform_stack_1.TerraformStack.of(this);
    }
    /**
     * @experimental
     */
    get constructNode() {
        return constructs_1.Node.of(this);
    }
    /**
     * @experimental
     */
    toTerraform() {
        return {};
    }
    /**
     * @experimental
     */
    get friendlyUniqueId() {
        if (this._logicalIdOverride) {
            return this._logicalIdOverride;
        }
        else {
            return this.cdktfStack.getLogicalId(this);
        }
    }
    /**
     * (experimental) Overrides the auto-generated logical ID with a specific ID.
     *
     * @param newLogicalId The new logical ID to use for this stack element.
     * @experimental
     */
    overrideLogicalId(newLogicalId) {
        this._logicalIdOverride = newLogicalId;
    }
    /**
     * @experimental
     */
    addOverride(path, value) {
        const parts = path.split('.');
        let curr = this.rawOverrides;
        while (parts.length > 1) {
            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
            const key = parts.shift();
            // if we can't recurse further or the previous value is not an
            // object overwrite it with an object.
            const isObject = curr[key] != null && typeof (curr[key]) === 'object' && !Array.isArray(curr[key]);
            if (!isObject) {
                curr[key] = {};
            }
            curr = curr[key];
        }
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        const lastKey = parts.shift();
        curr[lastKey] = value;
    }
    /**
     * @experimental
     */
    get constructNodeMetadata() {
        var _b;
        return {
            metadata: {
                path: this.constructNode.path,
                uniqueId: this.friendlyUniqueId,
                stackTrace: (_b = this.constructNode.metadata.find((e) => e.type === 'stacktrace')) === null || _b === void 0 ? void 0 : _b.trace
            }
        };
    }
}
exports.TerraformElement = TerraformElement;
_a = JSII_RTTI_SYMBOL_1;
TerraformElement[_a] = { fqn: "cdktf.TerraformElement", version: "0.4.1" };
