// cdktf/lib/terraform-module.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.TerraformModule = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const terraform_element_1 = require("./terraform-element");
const terraform_provider_1 = require("./terraform-provider");
const util_1 = require("./util");
const tokens_1 = require("./tokens");
const path = require("path");
/**
 * @experimental
 */
class TerraformModule extends terraform_element_1.TerraformElement {
    /**
     * @experimental
     */
    constructor(scope, id, options) {
        super(scope, id);
        if (options.source.startsWith('./') || options.source.startsWith('../')) {
            this.source = path.join('..', options.source);
        }
        else {
            this.source = options.source;
        }
        this.version = options.version;
        this._providers = options.providers;
        if (Array.isArray(options.dependsOn)) {
            this.dependsOn = options.dependsOn.map(dependency => dependency.fqn);
        }
    }
    // jsii can't handle abstract classes?
    /**
     * @experimental
     */
    synthesizeAttributes() {
        return {};
    }
    /**
     * @experimental
     */
    interpolationForOutput(moduleOutput) {
        return `\${module.${this.friendlyUniqueId}.${moduleOutput}}`;
    }
    /**
     * @experimental
     */
    get fqn() {
        return tokens_1.Token.asString(`module.${this.friendlyUniqueId}`);
    }
    /**
     * @experimental
     */
    get providers() {
        return this._providers;
    }
    /**
     * @experimental
     */
    addProvider(provider) {
        if (!this._providers) {
            this._providers = [];
        }
        this._providers.push(provider);
    }
    /**
     * @experimental
     */
    toTerraform() {
        var _b;
        const attributes = util_1.deepMerge({
            ...this.synthesizeAttributes(),
            source: this.source,
            version: this.version,
            providers: (_b = this.providers) === null || _b === void 0 ? void 0 : _b.map(p => {
                if (p instanceof terraform_provider_1.TerraformProvider) {
                    return { [p.terraformResourceType]: p.fqn };
                }
                else {
                    return { [`${p.provider.terraformResourceType}.${p.moduleAlias}`]: p.provider.fqn };
                }
            }),
            // eslint-disable-next-line @typescript-eslint/camelcase
            depends_on: this.dependsOn,
        }, this.rawOverrides);
        attributes['//'] = this.constructNodeMetadata;
        return {
            module: {
                [this.friendlyUniqueId]: attributes
            }
        };
    }
}
exports.TerraformModule = TerraformModule;
_a = JSII_RTTI_SYMBOL_1;
TerraformModule[_a] = { fqn: "cdktf.TerraformModule", version: "0.4.1" };
