// cdktf/lib/terraform-output.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.TerraformOutput = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const terraform_element_1 = require("./terraform-element");
const util_1 = require("./util");
/**
 * @experimental
 */
class TerraformOutput extends terraform_element_1.TerraformElement {
    /**
     * @experimental
     */
    constructor(scope, id, config) {
        super(scope, id);
        this.value = config.value;
        this.description = config.description;
        this.sensitive = config.sensitive;
        this.dependsOn = config.dependsOn;
    }
    /**
     * @experimental
     */
    synthesizeAttributes() {
        var _b;
        return {
            value: this.value,
            description: this.description,
            sensitive: this.sensitive,
            dependsOn: (_b = this.dependsOn) === null || _b === void 0 ? void 0 : _b.map(resource => `\${${resource.fqn}}`)
        };
    }
    /**
     * @experimental
     */
    toTerraform() {
        return {
            output: {
                [this.friendlyUniqueId]: util_1.deepMerge(util_1.keysToSnakeCase(this.synthesizeAttributes()), this.rawOverrides)
            }
        };
    }
}
exports.TerraformOutput = TerraformOutput;
_a = JSII_RTTI_SYMBOL_1;
TerraformOutput[_a] = { fqn: "cdktf.TerraformOutput", version: "0.4.1" };
