// cdktf/lib/terraform-provider.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.TerraformProvider = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const tokens_1 = require("./tokens");
const terraform_element_1 = require("./terraform-element");
const util_1 = require("./util");
/**
 * @experimental
 */
class TerraformProvider extends terraform_element_1.TerraformElement {
    /**
     * @experimental
     */
    constructor(scope, id, config) {
        super(scope, id);
        this.terraformResourceType = config.terraformResourceType;
        this.terraformGeneratorMetadata = config.terraformGeneratorMetadata;
        this.terraformProviderSource = config.terraformProviderSource;
    }
    /**
     * @experimental
     */
    get alias() {
        // This is always* being overriden currently
        return undefined;
    }
    /**
     * @experimental
     */
    set alias(_value) {
        // This is always* being overriden currently
    }
    /**
     * @experimental
     */
    get fqn() {
        return (this.alias !== undefined) ? tokens_1.Token.asString(`${this.terraformResourceType}.${this.alias}`) : tokens_1.Token.asString(`${this.terraformResourceType}`);
    }
    /**
     * @experimental
     */
    get metaAttributes() {
        return (this.alias !== undefined) ? { alias: this.alias } : {};
    }
    // jsii can't handle abstract classes?
    /**
     * @experimental
     */
    synthesizeAttributes() {
        return {};
    }
    /**
     * (experimental) Adds this resource to the terraform JSON output.
     *
     * @experimental
     */
    toTerraform() {
        var _b;
        return {
            terraform: {
                // eslint-disable-next-line @typescript-eslint/camelcase
                required_providers: {
                    [this.terraformResourceType]: {
                        version: (_b = this.terraformGeneratorMetadata) === null || _b === void 0 ? void 0 : _b.providerVersionConstraint,
                        source: this.terraformProviderSource
                    }
                }
            },
            provider: {
                [this.terraformResourceType]: [util_1.deepMerge(util_1.keysToSnakeCase(this.synthesizeAttributes()), this.rawOverrides, this.metaAttributes)]
            }
        };
    }
}
exports.TerraformProvider = TerraformProvider;
_a = JSII_RTTI_SYMBOL_1;
TerraformProvider[_a] = { fqn: "cdktf.TerraformProvider", version: "0.4.1" };
