// cdktf/lib/terraform-local.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.TerraformLocal = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const terraform_element_1 = require("./terraform-element");
const tokens_1 = require("./tokens");
/**
 * @experimental
 */
class TerraformLocal extends terraform_element_1.TerraformElement {
    /**
     * @experimental
     */
    constructor(scope, id, expression) {
        super(scope, id);
        this._expression = expression;
    }
    /**
     * @experimental
     */
    set expression(value) {
        this._expression = value;
    }
    /**
     * @experimental
     */
    get expression() {
        return tokens_1.Token.asAny(this.interpolation());
    }
    /**
     * @experimental
     */
    get asString() {
        return tokens_1.Token.asString(this.interpolation());
    }
    /**
     * @experimental
     */
    get asNumber() {
        return tokens_1.Token.asNumber(this.interpolation());
    }
    /**
     * @experimental
     */
    get asList() {
        return tokens_1.Token.asList(this.interpolation());
    }
    /**
     * @experimental
     */
    get asBoolean() {
        return tokens_1.Token.asString(this.interpolation());
    }
    interpolation() {
        return `\${local.${this.friendlyUniqueId}}`;
    }
    /**
     * @experimental
     */
    toTerraform() {
        return {
            locals: {
                [this.friendlyUniqueId]: this._expression
            }
        };
    }
}
exports.TerraformLocal = TerraformLocal;
_a = JSII_RTTI_SYMBOL_1;
TerraformLocal[_a] = { fqn: "cdktf.TerraformLocal", version: "0.4.1" };
