// @cdktf/provider-docker/lib/registry-image.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.RegistryImage = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const cdktf = require("cdktf");
function registryImageBuildAuthConfigToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    return {
        auth: cdktf.stringToTerraform(struct.auth),
        email: cdktf.stringToTerraform(struct.email),
        host_name: cdktf.stringToTerraform(struct.hostName),
        identity_token: cdktf.stringToTerraform(struct.identityToken),
        password: cdktf.stringToTerraform(struct.password),
        registry_token: cdktf.stringToTerraform(struct.registryToken),
        server_address: cdktf.stringToTerraform(struct.serverAddress),
        user_name: cdktf.stringToTerraform(struct.userName),
    };
}
function registryImageBuildUlimitToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    return {
        hard: cdktf.numberToTerraform(struct.hard),
        name: cdktf.stringToTerraform(struct.name),
        soft: cdktf.numberToTerraform(struct.soft),
    };
}
function registryImageBuildToTerraform(struct) {
    if (!cdktf.canInspect(struct)) {
        return struct;
    }
    return {
        build_args: cdktf.hashMapper(cdktf.anyToTerraform)(struct.buildArgs),
        build_id: cdktf.stringToTerraform(struct.buildId),
        cache_from: cdktf.listMapper(cdktf.stringToTerraform)(struct.cacheFrom),
        cgroup_parent: cdktf.stringToTerraform(struct.cgroupParent),
        context: cdktf.stringToTerraform(struct.context),
        cpu_period: cdktf.numberToTerraform(struct.cpuPeriod),
        cpu_quota: cdktf.numberToTerraform(struct.cpuQuota),
        cpu_set_cpus: cdktf.stringToTerraform(struct.cpuSetCpus),
        cpu_set_mems: cdktf.stringToTerraform(struct.cpuSetMems),
        cpu_shares: cdktf.numberToTerraform(struct.cpuShares),
        dockerfile: cdktf.stringToTerraform(struct.dockerfile),
        extra_hosts: cdktf.listMapper(cdktf.stringToTerraform)(struct.extraHosts),
        force_remove: cdktf.booleanToTerraform(struct.forceRemove),
        isolation: cdktf.stringToTerraform(struct.isolation),
        labels: cdktf.hashMapper(cdktf.anyToTerraform)(struct.labels),
        memory: cdktf.numberToTerraform(struct.memory),
        memory_swap: cdktf.numberToTerraform(struct.memorySwap),
        network_mode: cdktf.stringToTerraform(struct.networkMode),
        no_cache: cdktf.booleanToTerraform(struct.noCache),
        platform: cdktf.stringToTerraform(struct.platform),
        pull_parent: cdktf.booleanToTerraform(struct.pullParent),
        remote_context: cdktf.stringToTerraform(struct.remoteContext),
        remove: cdktf.booleanToTerraform(struct.remove),
        security_opt: cdktf.listMapper(cdktf.stringToTerraform)(struct.securityOpt),
        session_id: cdktf.stringToTerraform(struct.sessionId),
        shm_size: cdktf.numberToTerraform(struct.shmSize),
        squash: cdktf.booleanToTerraform(struct.squash),
        suppress_output: cdktf.booleanToTerraform(struct.suppressOutput),
        target: cdktf.stringToTerraform(struct.target),
        version: cdktf.stringToTerraform(struct.version),
        auth_config: cdktf.listMapper(registryImageBuildAuthConfigToTerraform)(struct.authConfig),
        ulimit: cdktf.listMapper(registryImageBuildUlimitToTerraform)(struct.ulimit),
    };
}
/**
 * Represents a {@link https://www.terraform.io/docs/providers/docker/r/registry_image.html docker_registry_image}.
 *
 * @stability stable
 */
class RegistryImage extends cdktf.TerraformResource {
    // ===========
    // INITIALIZER
    // ===========
    /**
     * Create a new {@link https://www.terraform.io/docs/providers/docker/r/registry_image.html docker_registry_image} Resource.
     *
     * @param scope The scope in which to define this construct.
     * @param id The scoped construct ID.
     * @stability stable
     */
    constructor(scope, id, config) {
        super(scope, id, {
            terraformResourceType: 'docker_registry_image',
            terraformGeneratorMetadata: {
                providerName: 'docker'
            },
            provider: config.provider,
            dependsOn: config.dependsOn,
            count: config.count,
            lifecycle: config.lifecycle
        });
        this._keepRemotely = config.keepRemotely;
        this._name = config.name;
        this._build = config.buildAttribute;
    }
    // ==========
    // ATTRIBUTES
    // ==========
    // id - computed: true, optional: true, required: false
    /**
     * @stability stable
     */
    get id() {
        return this.getStringAttribute('id');
    }
    /**
     * @stability stable
     */
    get keepRemotely() {
        return this.getBooleanAttribute('keep_remotely');
    }
    /**
     * @stability stable
     */
    set keepRemotely(value) {
        this._keepRemotely = value;
    }
    /**
     * @stability stable
     */
    resetKeepRemotely() {
        this._keepRemotely = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get keepRemotelyInput() {
        return this._keepRemotely;
    }
    /**
     * @stability stable
     */
    get name() {
        return this.getStringAttribute('name');
    }
    /**
     * @stability stable
     */
    set name(value) {
        this._name = value;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get nameInput() {
        return this._name;
    }
    // sha256_digest - computed: true, optional: false, required: false
    /**
     * @stability stable
     */
    get sha256Digest() {
        return this.getStringAttribute('sha256_digest');
    }
    /**
     * @stability stable
     */
    get buildAttribute() {
        return this.interpolationForAttribute('build');
    }
    /**
     * @stability stable
     */
    set buildAttribute(value) {
        this._build = value;
    }
    /**
     * @stability stable
     */
    resetBuildAttribute() {
        this._build = undefined;
    }
    // Temporarily expose input value. Use with caution.
    /**
     * @stability stable
     */
    get buildAttributeInput() {
        return this._build;
    }
    // =========
    // SYNTHESIS
    // =========
    /**
     * @stability stable
     */
    synthesizeAttributes() {
        return {
            keep_remotely: cdktf.booleanToTerraform(this._keepRemotely),
            name: cdktf.stringToTerraform(this._name),
            build: cdktf.listMapper(registryImageBuildToTerraform)(this._build),
        };
    }
}
exports.RegistryImage = RegistryImage;
_a = JSII_RTTI_SYMBOL_1;
RegistryImage[_a] = { fqn: "@cdktf/provider-docker.RegistryImage", version: "0.2.31" };
